File: core/src/de/project/ice/utils/SceneWriter.java
package de.project.ice.utils;

import com.badlogic.ashley.core.Component;
import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.IntMap;
import com.badlogic.gdx.utils.XmlWriter;
import com.badlogic.gdx.utils.reflect.ClassReflection;
import com.badlogic.gdx.utils.reflect.Field;
import com.badlogic.gdx.utils.reflect.ReflectionException;
import de.project.ice.ecs.IceEngine;
import de.project.ice.ecs.systems.SoundSystem;
import de.project.ice.pathlib.PathArea;
import de.project.ice.pathlib.Shape;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public abstract class SceneWriter {
    public static void serializeScene(String sceneName, @NotNull IceEngine engine, @NotNull XmlWriter xml) throws IOException {
        xml.element("scene")
                .attribute("name", sceneName);


        serializeAudio(xml, engine.soundSystem);

        xml.element("entities");
        for (Entity entity : engine.getEntities()) {
            xml.element("entity");
            xml.element("components");
            for (Component component : entity.getComponents()) {
                xml.element(component.getClass().getSimpleName());
                for (Field f : ClassReflection.getFields(component.getClass())) {
                    if (f.isFinal())
                        continue;

                    xml.element(f.getName());
                    try {
                        serializeObject(xml, f.get(component));
                    } catch (ReflectionException e) {
                        e.printStackTrace();
                    }
                    xml.pop();
                }
                xml.pop();
            }
            xml.pop();
            xml.pop();
        }

        xml.pop();
        xml.pop();
        xml.close();
    }

    private static void serializeAudio(XmlWriter xml, SoundSystem sound) throws IOException {
        xml.element("music").text(sound.getMusic()).pop();

        xml.element("sounds");

        for(String s:sound.getSounds()){
            xml.element("sound").text(s).pop();
        }

        xml.pop();

    }

    private static void serializeObject(XmlWriter xml, Object o) throws IOException {
        if (o == null) return;
        Class type = o.getClass();
        if (Float.class.equals(type) || type == float.class ||
                Double.class.equals(type) || type == double.class ||
                Short.class.equals(type) || type == short.class ||
                Integer.class.equals(type) || type == int.class ||
                Long.class.equals(type) || type == long.class) {
            xml.attribute("type", type.getSimpleName());
            xml.text(o);
        } else if (String.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            xml.text(((String) o).replace("&", "&amp").replace("<", "&lt").replace(">", "&gt"));
        } else if (Boolean.class.equals(type) || type == boolean.class) {
            xml.attribute("type", type.getSimpleName());
            Boolean bool = (Boolean) o;
            xml.write(bool ? "true" : "false");
        } else if (Vector2.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            Vector2 vec = (Vector2) o;
            xml.attribute("x", vec.x);
            xml.attribute("y", vec.y);
        } else if (Vector3.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            Vector3 vec = (Vector3) o;
            xml.attribute("x", vec.x);
            xml.attribute("y", vec.y);
            xml.attribute("z", vec.z);
        } else if (Assets.AnimationHolder.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            Assets.AnimationHolder holder = (Assets.AnimationHolder) o;
            Animation animation = holder.data;
            xml.attribute("frameDuration", animation.getFrameDuration());
            xml.attribute("mode", animation.getPlayMode());
            xml.text(holder.name);
        } else if (Assets.TextureRegionHolder.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            Assets.TextureRegionHolder holder = (Assets.TextureRegionHolder) o;
            xml.text(holder.name);
        } else if (IntMap.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            for (IntMap.Entry<Object> entry : (IntMap<Object>) o) {
                if (entry.value != null) {
                    xml.attribute("subType", entry.value.getClass().getSimpleName());
                    break;
                }
            }
            for (IntMap.Entry<Object> entry : (IntMap<Object>) o) {
                xml.element("_" + entry.key);
                serializeObject(xml, entry.value);
                xml.pop();
            }
        } else if (OrthographicCamera.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            OrthographicCamera cam = (OrthographicCamera) o;
            xml.attribute("viewportWidth", cam.viewportWidth);
            xml.attribute("viewportHeight", cam.viewportHeight);
            xml.element("pos");
            serializeObject(xml, cam.position);
            xml.pop();
        } else if (PathArea.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            PathArea pathArea = (PathArea) o;

            xml.element("shape");
            serializeObject(xml, pathArea.shape);
            xml.pop();

            xml.element("holes");
            for (Shape hole : pathArea.holes) {
                xml.element("hole");
                serializeObject(xml, hole);
                xml.pop();
            }
            xml.pop();

        } else if (Shape.class.equals(type)) {
            xml.attribute("type", type.getSimpleName());
            Shape shape = (Shape) o;
            xml.attribute("closed", shape.closed);
            xml.element("vertices");
            for (Vector2 v : shape.vertices) {
                xml.element("vertex");
                serializeObject(xml, v);
                xml.pop();
            }
            xml.pop();
        } else {
            xml.attribute("type", "null");
        }
    }
}

File: core/src/de/project/ice/utils/SceneWriter.kt
package de.project.ice.utils

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.IntMap
import com.badlogic.gdx.utils.XmlWriter
import com.badlogic.gdx.utils.reflect.ReflectionException
import de.project.ice.ecs.IceEngine
import de.project.ice.ecs.systems.SoundSystem
import de.project.ice.pathlib.PathArea
import de.project.ice.pathlib.Shape
import java.io.IOException
import kotlin.reflect.jvm.javaField
import kotlin.reflect.memberProperties

class SceneWriter private constructor(private val sceneName: String, private val onloadScript: String, private val engine: IceEngine, private val xml: XmlWriter) {

    @Throws(IOException::class)
    fun serializeScene() {
        xml.element("scene").attribute("name", sceneName).attribute("onload", onloadScript)


        serializeAudio(xml, engine.soundSystem)
        serializeSpritesheets(xml)

        xml.element("entities")
        for (entity in engine.entities) {
            xml.element("entity")
            xml.element("components")
            for (component in entity.components) {
                xml.element(component.javaClass.simpleName)
                for (property in component.javaClass.kotlin.memberProperties) {
                    val f = property.javaField;

                    if (f == null) {
                        Gdx.app.log("SceneWriter", "No baking field found for property ${property.name} of component ${component.javaClass.simpleName}")
                        continue;
                    }

                    f.isAccessible = true

                    val value = f.get(component) ?: continue

                    xml.element(f.name)
                    try {
                        serializeObject(xml, value)
                    } catch (e: ReflectionException) {
                        e.printStackTrace()
                    }

                    xml.pop()
                }
                xml.pop()
            }
            xml.pop()
            xml.pop()
        }

        xml.pop()
        xml.pop()
        xml.close()
    }

    @Throws(IOException::class)
    private fun serializeSpritesheets(xml: XmlWriter) {
        xml.element("spritesheets")

        for (s in Assets.getLoadedSpritesheets()) {
            xml.element("spritesheet").text(s).pop()
        }

        xml.pop()
    }

    @Throws(IOException::class)
    private fun serializeAudio(xml: XmlWriter, sound: SoundSystem) {
        xml.element("music").text(sound.music).pop()

        xml.element("sounds")

        for (s in sound.sounds) {
            xml.element("sound").text(s).pop()
        }

        xml.pop()

    }

    @Throws(IOException::class)
    private fun serializeObject(xml: XmlWriter, o: Any) {
        when (o) {
            is Float -> {
                xml.attribute("type", "Float");  xml.text(o)
            }
            is Double -> {
                xml.attribute("type", "Double");  xml.text(o)
            }
            is Short -> {
                xml.attribute("type", "Short");  xml.text(o)
            }
            is Int -> {
                xml.attribute("type", "Integer");  xml.text(o)
            }
            is Long -> {
                xml.attribute("type", "Long");  xml.text(o)
            }
            is String -> {
                xml.attribute("type", "String");  xml.text(o.replace("&", "&amp").replace("<", "&lt").replace(">", "&gt"))
            }
            is Boolean -> {
                xml.attribute("type", "Boolean");  xml.text(if (o) "true" else "false")
            }
            is Vector2 -> {
                xml.attribute("type", o.javaClass.simpleName);  xml.attribute("x", o.x).attribute("y", o.y)
            }
            is Vector3 -> {
                xml.attribute("type", o.javaClass.simpleName);  xml.attribute("x", o.x).attribute("y", o.y).attribute("z", o.z)
            }
            is Assets.Holder.Animation -> {
                xml.attribute("type", o.javaClass.simpleName)
                if (o.data != null) xml
                        .attribute("frameDuration", o.data.animationDuration)
                        .attribute("mode", o.data.playMode).text(o.name)
            }
            is Assets.Holder.TextureRegion -> {
                xml.attribute("type", o.javaClass.simpleName); xml.text(o.name);
            }
            is IntMap<*> -> {
                xml.attribute("type", o.javaClass.simpleName)
                for (entry in o) if (entry.value != null) {
                    xml.attribute("subType", entry.value.javaClass.simpleName)
                    break
                }
                for (entry in o) {
                    xml.element("_${entry.key}")
                    serializeObject(xml, entry.value)
                    xml.pop()
                }
            }
            is OrthographicCamera -> {
                xml.attribute("type", o.javaClass.simpleName)
                xml.attribute("viewportWidth", o.viewportWidth)
                        .attribute("viewportHeight", o.viewportHeight)
                        .element("pos")
                serializeObject(xml, o.position)
                xml.pop()
            }
            is PathArea -> {
                xml.attribute("type", o.javaClass.simpleName)
                xml.element("shape")
                serializeObject(xml, o.shape)
                xml.pop()

                xml.element("holes")
                for (hole in o.holes) {
                    xml.element("hole")
                    serializeObject(xml, hole)
                    xml.pop()
                }
                xml.pop()
            }
            is Shape -> {
                xml.attribute("type", o.javaClass.simpleName)
                xml.attribute("closed", o.closed)
                xml.element("vertices")
                for (v in o.vertices) {
                    xml.element("vertex")
                    serializeObject(xml, v)
                    xml.pop()
                }
                xml.pop()
            }
            else -> {
                xml.attribute("type", "null")
                return
            }
        }
    }

    class Builder {
        private var sceneName = ""
        private var onloadScript = ""
        private var engine: IceEngine? = null
        private var writer: XmlWriter? = null

        fun sceneName(sceneName: String): Builder {
            this.sceneName = sceneName
            return this
        }

        fun onloadScript(onloadScript: String): Builder {
            this.onloadScript = onloadScript
            return this
        }

        fun engine(engine: IceEngine): Builder {
            this.engine = engine
            return this
        }

        fun writer(writer: XmlWriter): Builder {
            this.writer = writer
            return this
        }

        fun create(): SceneWriter {
            if (engine == null || writer == null) {
                throw RuntimeException("Trying to create SceneWriter without XmlWriter or Engine")
            }
            return SceneWriter(sceneName, onloadScript, engine!!, writer!!)
        }
    }
}
